invokeai/backend/model_manager/config.py

```diff
@@ -0,0 +1,342 @@
+import warnings
+from enum import Enum
+from typing import List, Literal, Optional, Type, Union
+
+import pydantic
+
+# import these so that we can silence them
+from diffusers import logging as diffusers_logging
+from omegaconf.listconfig import ListConfig  # to support the yaml backend
+from pydantic import BaseModel, Extra, Field
+from pydantic.error_wrappers import ValidationError
+from transformers import logging as transformers_logging
+
+
+class InvalidModelConfigException(Exception):
+    """Exception for when config parser doesn't recognized this combination of model type and format."""
+
+
+class BaseModelType(str, Enum):
+    """Base model type."""
+
+    Any = "any"
+    StableDiffusion1 = "sd-1"
+    StableDiffusion2 = "sd-2"
+    StableDiffusionXL = "sdxl"
+    StableDiffusionXLRefiner = "sdxl-refiner"
+    # Kandinsky2_1 = "kandinsky-2.1"
+
+
+class ModelType(str, Enum):
+    """Model type."""
+
+    ONNX = "onnx"
+    Main = "main"
+    Vae = "vae"
+    Lora = "lora"
+    ControlNet = "controlnet"  # used by model_probe
+    TextualInversion = "embedding"
+    IPAdapter = "ip_adapter"
+    CLIPVision = "clip_vision"
+
+
+class SubModelType(str, Enum):
+    """Submodel type."""
+
+    UNet = "unet"
+    TextEncoder = "text_encoder"
+    TextEncoder2 = "text_encoder_2"
+    Tokenizer = "tokenizer"
+    Tokenizer2 = "tokenizer_2"
+    Vae = "vae"
+    VaeDecoder = "vae_decoder"
+    VaeEncoder = "vae_encoder"
+    Scheduler = "scheduler"
+    SafetyChecker = "safety_checker"
+
+
+class ModelVariantType(str, Enum):
+    """Variant type."""
+
+    Normal = "normal"
+    Inpaint = "inpaint"
+    Depth = "depth"
+
+
+class ModelFormat(str, Enum):
+    """Storage format of model."""
+
+    Diffusers = "diffusers"
+    Checkpoint = "checkpoint"
+    Lycoris = "lycoris"
+    Onnx = "onnx"
+    Olive = "olive"
+    EmbeddingFile = "embedding_file"
+    EmbeddingFolder = "embedding_folder"
+    InvokeAI = "invokeai"
+
+
+class SchedulerPredictionType(str, Enum):
+    """Scheduler prediction type."""
+
+    Epsilon = "epsilon"
+    VPrediction = "v_prediction"
+    Sample = "sample"
+
+
+# TODO: use this
+class ModelError(str, Enum):
+    NotFound = "not_found"
+
+
+class ModelConfigBase(BaseModel):
+    """Base class for model configuration information."""
+
+    path: str
+    name: str
+    base_model: BaseModelType
+    model_type: ModelType
+    model_format: ModelFormat
+    key: str = Field(description="hash key for model", default="<NOKEY>")  # this will get added by the store
+    description: Optional[str] = Field(None)
+    author: Optional[str] = Field(description="Model author")
+    license: Optional[str] = Field(description="License string")
+    source: Optional[str] = Field(description="Model download source (URL or repo_id)")
+    thumbnail_url: Optional[str] = Field(description="URL of thumbnail image")
+    tags: Optional[List[str]] = Field(description="Descriptive tags")  # Set would be better, but not JSON serializable
```

Comment:
I appreciate these extra bits of metadata and the effort needed to implement them all, but I'm not sure if we want to include them just yet.

Can we keep this minimal for now, and only include the essential data? That'd be the required fields, and *maybe* `source`.

Tags in particular probably need some consideration - there are many ways to implement them.

Reply:
What's the argument to remove them? They are already automatically populated when the user installs a model using either a huggingface repo_id or a civitai url. 

Tags are implemented in a normalized fashion in the sqlite database.

Reply:
That sounds like we are caching hf/civitai server state locally. That's *far* more complex and than just storing the URL and having the client query the hf/civitai API to get these extra details when needed. It's a lot of extra responsibility for zero benefit (at this time).

If we treat this data as a cache, then we need to invalidate it so that it stays in sync with the source. If this data is intended to be modified by the user, do we let them just edit the existing information? Replace it or add to it?

We won't have a UI for this for some time, so it's just unused code to get broken. The first release of this branch will not include any big UI changes, it's just getting the backend working and ironing out any kinks.

Re: tags, there's a clear path for tags for models but what about boards, images, workflows, and who knows what else? Let's add app-wide tags in a dedicated feature release.